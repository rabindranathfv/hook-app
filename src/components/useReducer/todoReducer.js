const ADD = 'add';
const DEL ='delete';
const TOG = 'toggle';
const TOG_OLD = 'toggle-old';

export const todoReducer = ( state = [], action ) => {

    switch ( action.type ) {
        case ADD:
            return [ ...state, action.payload ];
        case DEL:
            return state.filter( todo => todo.id !== action.payload );
        case TOG: 
            return state.map( todo => 
                ( todo.id === action.payload )
                    ? { ...todo, done: !todo.done }
                    : todo
            );
        case TOG_OLD:
            return state.map( todo => {

                if ( todo.id === action.payload ) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }
            })
        default:
            return state;
    }


}