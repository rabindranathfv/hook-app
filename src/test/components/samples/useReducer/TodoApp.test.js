import React from 'react';
import { shallow } from 'enzyme';
import { TodoApp } from './../../../../components/useReducer/TodoApp';

const todosMock = [
    { id: 1, desc: 't1', done: false},
    { id: 2, desc: 't2', done: false},
    { id: 3, desc: 't3', done: true }
];

describe('TodoApp', () => {

    let wrapper = shallow(<TodoApp />);

    beforeEach(() => {
        localStorage.setItem('todos', JSON.stringify(todosMock));
        wrapper = shallow(<TodoApp />);
    })
    it('Should render TodoApp component', async () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('Should render all todos from localStorage and todoList component', () => {
        const todoList = wrapper.find('TodoList');
        todoList.debug();
        console.log('***TODO ', todoList.props());

        expect(todoList.props().todos.length).toBe(3);
        expect(typeof todoList.props().handleDelete).toBe('function');
        expect(typeof todoList.props().handleToggle).toBe('function');
    });

    it('Should render todoAddComponent', () => {
        const todoAdd = wrapper.find('TodoAdd');
        todoAdd.debug();
        console.log('***TODO ', todoAdd.props());

        expect(typeof todoAdd.props().handleAddTodo).toBe('function');
    });
})