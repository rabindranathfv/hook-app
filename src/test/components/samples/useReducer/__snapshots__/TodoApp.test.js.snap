// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp Should render TodoApp component 1`] = `
<div>
  <h1>
    TodoApp ( 
    3
     ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={
          Array [
            Object {
              "desc": "t1",
              "done": false,
              "id": 1,
            },
            Object {
              "desc": "t2",
              "done": false,
              "id": 2,
            },
            Object {
              "desc": "t3",
              "done": true,
              "id": 3,
            },
          ]
        }
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
